symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Optimization.hs

```
@@ -0,0 +1,98 @@
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Basic.Number
+import           ZkFold.Base.Algebra.Polynomials.Multivariate            (evalMonomial)
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Monomial   (Mono (..), oneM)
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Polynomial (Poly (..), evalPolynomial, var)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Instance     ()
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.MerkleHash   (merkleHash, runHash)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Witness      (WitnessF (..))
+
+--------------------------------- High-level functions --------------------------------
+
+-- | Replaces linear polynoms of the form
+-- @(fromConstant k) * (NewVar nV) + (fromConstant c)@
+-- with a constant variable @fromConstant $ negate c // k@ in an arithmetic circuit
+-- and replaces variable with a constant in witness
+--
+optimize ::
+   (Arithmetic a, Ord (Rep i), Functor o, Binary (Rep i)) =>
+  ArithmeticCircuit a p i o  -> ArithmeticCircuit a p i o
+optimize ac = let (newAcSystem, vs) = varsToReplace (acSystem ac, M.empty) in optimizeAc vs ac {acSystem = newAcSystem}
+
+varsToReplace ::
+  (Arithmetic a, Ord (Rep i)) =>
+  (Map ByteString (Constraint a i) , Map (SysVar i) a) -> (Map ByteString (Constraint a i) , Map (SysVar i) a)
+varsToReplace (s, l) = let newVars = M.fromList . M.elems $ mapMaybe toConstVar s in
+  if newVars == M.empty
+  then (s, l)
+  else varsToReplace (optimizeSystems newVars s, M.union newVars l)
+
+optimizeSystems :: (Arithmetic a, Ord (Rep i)) =>
+  Map (SysVar i) a -> Map ByteString (Constraint a i) -> Map ByteString (Constraint a i)
+optimizeSystems m as = M.filter (/= zero) $ evalPolynomial evalMonomial varF <$> as
```

Same as range checks, maybe we want to throw an error here if we encounter an unsatisfiable constraint?